### DISCUSSION TITLE
Run FSAC inside a Docker container

### DISCUSSION THREAD

**Discussion by nojaf** [ionide/ionide-vscode-fsharp#1494] [Ideas]
Hi, I'm wondering if it would be possible for the ionide extension to connect to a running Docker container with .NET in it.
So imagine I spin a container and keep it running in the background:
```
docker run -t -d --rm --name dotnet-container mcr.microsoft.com/dotnet/sdk:5.0-focal
```

I map the volumes and such and then the FASC is started from there.
Expose the right ports and connection from the editor.
Would such a thing be feasible?

**Comment by baronfel** [comment]
I think it _could_ be possible.  VSCode lets us spawn a LanguageServer and interact with it via a type called `ServerOptions`, which is defined [here](https://github.com/microsoft/vscode-languageserver-node/blob/master/client/src/node/main.ts#L124).

```typescript
export type ServerOptions = Executable | { run: Executable; debug: Executable; } | { run: NodeModule; debug: NodeModule } | NodeModule | (() => Promise<ChildProcess | StreamInfo | MessageTransports | ChildProcessInfo>);
```

Of all of these options, I think it would be pretty simple to return a `StreamInfo` here:

```typescript
export interface StreamInfo {
	writer: NodeJS.WritableStream;
	reader: NodeJS.ReadableStream;
	detached?: boolean;
}
```

where the WritableStream would be mapped to the `stdin` of your container, and the Readable stream would be mapped to the `stdout` of your container.  I think if you can get the stdin/stdout of the external process somehow that might also work.

**Reply by nojaf** [reply]
Would it not be easier to "connect" to a language server instead of spawning one?
I'm going to make some assumptions but:
- FSAC runs some sort of webserver to deal with all the LSP protocols?
- Could I just not then connect to the existing running server instead of spawning one myself?
- The fact that I'm going to run FSAC inside a docker container should be transparent for the extension, as in it does not need to know right?

**Reply by baronfel** [reply]
That's what I'm saying: the mechanism to connect to an already running server is to provide vscode with a pair of stream representing the input to that server and the output from that server.

You're not necessarily spawning it, spawning a new server is just one of the options (most of the other serveroptions in the union I mentioned deal with explicitly spawning new  processes).

**Reply by nojaf** [reply]
Ok, I had a different idea of how that part worked.
Would it be as "simple" as extending https://github.com/ionide/ionide-vscode-fsharp/blob/9cad0e4fc969bfc99d01a80f6024bbebe85a194a/src/Core/LanguageService.fs#L648-L692 ?

**Comment by Krzysztof-Cieslak** [comment]
Have you tried dev containers? (https://code.visualstudio.com/docs/remote/containers)

**Reply by nojaf** [reply]
I'm playing around in gitpod primarily. The simple answer would be to just use their .NET image and install the extensions from openvisx.

However, ever since you can run containers inside a workspace ([see blogpost](https://www.gitpod.io/blog/root-docker-and-vscode/#docker)), I'm sorta exploring the workflow of having everything inside containers and no dependency on the base image.

See the [sample repository](https://github.com/nojaf/fs-advent-2020#022021-revision).

**Reply by Krzysztof-Cieslak** [reply]
You may try using dev containers to connect to docker running inside gitpod instance.

**Reply by nojaf** [reply]
I'm using Theia editor at the moment (as there is current is a bug with [vscode in gitpod](https://github.com/gitpod-io/gitpod/issues/3114)) so I cannot check whether dev containers are even a thing there.
This honestly seems like a very tied to vscode type of functionality.

**Reply by Krzysztof-Cieslak** [reply]
It's not tied to vscode any more than Ionide itself, it's just an extension. It should probably run fine in gitpod/theia 🤷‍♂️

**Reply by nojaf** [reply]
Do you have any proof of that? The branding on the [docs ](https://code.visualstudio.com/docs/remote/containers) is very on the nose here:
> The Visual Studio Code Remote - Containers extension lets you use a Docker container as a full-featured development environment.

It literally is a extension provided by MSFT: https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers
Nowhere to be found on https://open-vsx.org/

**Reply by Krzysztof-Cieslak** [reply]
Ionide up until yesterday was also not found on open-vsx. And yet it worked fine on GitPod for years.
